### ISSUE TITLE
[v2 BUG]: Bug in calculating the initial and final learning rates while using hpopt

### ISSUE THREAD

**Issue by akshatzalte** [chemprop/chemprop#912]
**Describe the bug**
hpopt finds the best init-lr-ratio, final-lr-ratio which are not directly consistent with the training arguments. #881 fixes that. 
The init-lr and final-lr are calculated using the default max-lr instead of the best max-lr calculated by hpopt.

**Example(s)**
the best_config.toml file:
```smiles-columns = [resonance_smis]
num-workers = 20
keep-h = true
add-h = true
data-path = /home/akshatz/bond_order_free/rigr_h298_50k/dataset/run1/data_run_1.csv
message-hidden-dim = 300
depth = 5
aggregation = sum
ffn-hidden-dim = 1200
ffn-num-layers = 2
target-columns = [H298_kcal]
task-type = regression
epochs = 100
splits-file = /home/akshatz/bond_order_free/rigr_h298_50k/dataset/run1/splits.json
batch-size = 64
accelerator = auto
devices = auto
rxn-mode = REAC_DIFF
multi-hot-atom-featurizer-mode = V2
frzn-ffn-layers = 0
ensemble-size = 1
dropout = 0.0
activation = RELU
aggregation-norm = 100
multiclass-num-classes = 3
v-kl = 0.0
eps = 1e-08
warmup-epochs = 48
init-lr = 4.989663651419804e-05
max-lr = 0.00039831600062360307
final-lr = 2.0161614463867722e-05
split = RANDOM
split-sizes = [0.8, 0.1, 0.1]
split-key-molecule = 0
num-folds = 1
data-seed = 0```

Corresponding all_progress.csv (only best row):
final_lr_ratio	warmup_epochs	init_lr_ratio	max_lr
0.02016161446	48	0.04989663651	0.0003983160006


**Expected behavior**
best max-lr to be used instead of default max-lr to get the best init-lr and final-lr from the the best linit-lr-ratio, final-lr-ratio and max-lr outputted from hpopt.

**Error Stack Trace**
If the bug is resulting in an error message, provide the _full_ stack trace (not just the last line). This is helpful for debugging, especially in cases where you aren't able to provide a minimum/isolated working example with accompanying files.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment**
- python version
- package versions: `conda list` or `pip list`
- OS

**Checklist**
- [ ] all dependencies are satisifed: `conda list` or `pip list` shows the packages listed in the `pyproject.toml`
- [ ] the unit tests are working: `pytest -v` reports no errors

**Additional context**
Add any other context about the problem here.

**Comment by shihchengli** [comment]
I was wondering if this also affects the init_lr and final_lr for each trial in HPO, as `update_args_with_config` is also called in `train_model`. If this is the case, then the init_lr and final_lr saved in best_config.toml represent the best configuration found by HPO. However, we might encounter situations where the max_lr is lower than either the init_lr or the final_lr.

**Comment by akshatzalte** [comment]
For each trial, the set of warmup-epochs, init-lr-ratio, final-lr-ratio and max-lr is used. The issue here is just concerning the conversion of the init-lr-ratio and final-lr-ratio to init-lr and final-lr respectively (which should be done using the max-lr for that trial itself) Currently, this conversion is done using the default max-lr of 0.001 if I recall correctly which might lead to getting the init-lr and final-lr to be greater than max-lr which would be wrong (as no such set of values would be or was present in the hpopt trial)

**Comment by shihchengli** [comment]
I agree with most of it, but I think the last sentence might be wrong. The conversion of the init-lr-ratio and final-lr-ratio to init-lr and final-lr happens for each trial and the final step to save them into a .toml file. Therefore, I think that the set of values in the .toml file are those that were present in the hpopt trial. Of course, we don't want max-lr to be lower than either the init-lr or the final-lr. I think this approach would lead to poor model performance, and thus it would not be the best trial in HPO. As a result, it would not be saved into the configuration file.

**Comment by akshatzalte** [comment]
The set of values used in hpopt cannot have max-lr smaller than init-lr (or final-lr) by definition.
hpopt doesn't use init-lr (or final-lr) directly, it uses the init-lr-ratio and final-lr-ratio (whose search space is limited to values less than 1), so none of the hpopt trials can have max-lr smaller than init-lr as the ratio = init-lr/max-lr < 1 (always!).
The .toml file due to the bug did this calculation of the init-lr erroneously leading to the artifact you mentioned.
